pkg/config/config.go

```diff
@@ -70,257 +74,255 @@ type Proxy struct {
 func init() {
 	// Configure Datadog global configuration
 	Datadog = NewConfig("datadog", "DD", strings.NewReplacer(".", "_"))
-
 	// Configuration defaults
+	initConfig(Datadog)
+}
+
+// initConfig initializes the config defaults on a config
+func initConfig(config Config) {
 	// Agent
-	BindEnvAndSetDefault("dd_url", "https://app.datadoghq.com")
-	BindEnvAndSetDefault("app_key", "")
-	Datadog.SetDefault("proxy", nil)
-	BindEnvAndSetDefault("skip_ssl_validation", false)
-	BindEnvAndSetDefault("hostname", "")
-	BindEnvAndSetDefault("tags", []string{})
-	BindEnvAndSetDefault("tag_value_split_separator", map[string]string{})
-	BindEnvAndSetDefault("conf_path", ".")
-	BindEnvAndSetDefault("confd_path", defaultConfdPath)
-	BindEnvAndSetDefault("additional_checksd", defaultAdditionalChecksPath)
-	BindEnvAndSetDefault("log_payloads", false)
-	BindEnvAndSetDefault("log_file", "")
-	BindEnvAndSetDefault("log_level", "info")
-	BindEnvAndSetDefault("log_to_syslog", false)
-	BindEnvAndSetDefault("log_to_console", true)
-	BindEnvAndSetDefault("logging_frequency", int64(20))
-	BindEnvAndSetDefault("disable_file_logging", false)
-	BindEnvAndSetDefault("syslog_uri", "")
-	BindEnvAndSetDefault("syslog_rfc", false)
-	BindEnvAndSetDefault("syslog_pem", "")
-	BindEnvAndSetDefault("syslog_key", "")
-	BindEnvAndSetDefault("syslog_tls_verify", true)
-	BindEnvAndSetDefault("cmd_host", "localhost")
-	BindEnvAndSetDefault("cmd_port", 5001)
-	BindEnvAndSetDefault("cluster_agent.cmd_port", 5005)
-	BindEnvAndSetDefault("default_integration_http_timeout", 9)
-	BindEnvAndSetDefault("enable_metadata_collection", true)
-	BindEnvAndSetDefault("enable_gohai", true)
-	BindEnvAndSetDefault("check_runners", int64(4))
-	BindEnvAndSetDefault("auth_token_file_path", "")
-	BindEnvAndSetDefault("bind_host", "localhost")
+	// Don't set a default on 'site' to allow detecting with viper whether it's set in config
+	config.BindEnv("site")
```

Comment:
Should we have a default for site ? Like we had for dd_url ?
What if people don't have `dd_url` in their configuration and upgrade to a new version of the agent ?

Reply:
We can't set the default of `site` here because of the comment above (`to allow detecting with viper whether it's set in config`) . Same reason for `dd_url` actually (I'll update the comment).

I _think_ the PR description (that I updated slightly) answers your last question:

>`dd_url` is not necessarily set on `config.Datadog` anymore. The "resolved" URL for the infra Agent should now be accessed with `config.GetMainInfraEndpoint`.

So if users don't have `dd_url` set in their config, `config.Datadog.GetString("dd_url")` will now return an empty value, but the Agent will still use all the expected defaults because it now accesses the default endpoint with `config.Datadog.GetMainInfraEndpoint`.

Reply:
Why move the default logic from the viper to the code ?

We could add a default to `site` and remove the elseif/else here https://github.com/DataDog/datadog-agent/pull/2412/files#diff-091ace02b49a46f55f2586cfec89a160R513

What do you think ?

Reply:
if we set a default with Viper directly, we can't know anymore whether the user actually set the option or not in their config/env vars (see convo there: https://github.com/DataDog/datadog-agent/pull/2412#discussion_r222917900). Since we want to display a log message when `site` and `dd_url` are both set, we need to know whether the user actually set these settings

Reply:
I see, ok :+1: